File: DeviceIDRotatorDylib/DeviceIDRotator.m
//
//  DeviceIDRotator.m
//  Device ID Rotator Dylib
//  Hooks into UIDevice to provide custom device ID rotation
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import <objc/runtime.h>

static NSString *const kCustomDeviceIDKey = @"com.swiggy.customDeviceID";
static NSString *const kDeviceIDRotatorNotification = @"com.swiggy.deviceIDRotated";

// Original method implementations
static id (*original_identifierForVendor)(id, SEL);
static NSString *(*original_advertisingIdentifier)(id, SEL);

@interface DeviceIDRotator : NSObject
+ (NSString *)getCurrentDeviceID;
+ (NSString *)rotateDeviceID;
+ (NSString *)generateNewDeviceID;
@end

@implementation DeviceIDRotator

+ (NSString *)getCurrentDeviceID {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *storedID = [defaults stringForKey:kCustomDeviceIDKey];
    
    if (!storedID) {
        storedID = [self generateNewDeviceID];
    }
    
    return storedID;
}

+ (NSString *)generateNewDeviceID {
    NSString *newID = [[NSUUID UUID] UUIDString];
    [[NSUserDefaults standardUserDefaults] setObject:newID forKey:kCustomDeviceIDKey];
    [[NSUserDefaults standardUserDefaults] synchronize];
    
    // Post notification
    [[NSNotificationCenter defaultCenter] postNotificationName:kDeviceIDRotatorNotification 
                                                        object:newID];
    
    return newID;
}

+ (NSString *)rotateDeviceID {
    return [self generateNewDeviceID];
}

@end

// Hook for UIDevice.identifierForVendor
static NSUUID *swizzled_identifierForVendor(id self, SEL _cmd) {
    NSString *customID = [DeviceIDRotator getCurrentDeviceID];
    NSUUID *uuid = [[NSUUID alloc] initWithUUIDString:customID];
    return uuid;
}

// Hook for ASIdentifierManager (if available)
static NSString *swizzled_advertisingIdentifier(id self, SEL _cmd) {
    return [DeviceIDRotator getCurrentDeviceID];
}

// Constructor - runs when dylib is loaded
__attribute__((constructor))
static void init() {
    @autoreleasepool {
        NSLog(@"[DeviceIDRotator] Dylib loaded successfully");
        
        // Swizzle UIDevice.identifierForVendor
        Class deviceClass = objc_getClass("UIDevice");
        if (deviceClass) {
            Method originalMethod = class_getInstanceMethod(deviceClass, @selector(identifierForVendor));
            if (originalMethod) {
                original_identifierForVendor = (void *)method_getImplementation(originalMethod);
                method_setImplementation(originalMethod, (IMP)swizzled_identifierForVendor);
                NSLog(@"[DeviceIDRotator] Hooked UIDevice.identifierForVendor");
            }
        }
        
        // Try to hook ASIdentifierManager (for advertising identifier)
        Class asIdentifierClass = objc_getClass("ASIdentifierManager");
        if (asIdentifierClass) {
            Method adMethod = class_getInstanceMethod(asIdentifierClass, @selector(advertisingIdentifier));
            if (adMethod) {
                original_advertisingIdentifier = (void *)method_getImplementation(adMethod);
                method_setImplementation(adMethod, (IMP)swizzled_advertisingIdentifier);
                NSLog(@"[DeviceIDRotator] Hooked ASIdentifierManager.advertisingIdentifier");
            }
        }
        
        // Initialize with a device ID if none exists
        [DeviceIDRotator getCurrentDeviceID];
        
        NSLog(@"[DeviceIDRotator] Initialization complete. Current Device ID: %@", 
              [DeviceIDRotator getCurrentDeviceID]);
    }
}

// Export functions for external access
void rotateDeviceID(void) {
    [DeviceIDRotator rotateDeviceID];
}

const char* getCurrentDeviceID(void) {
    NSString *deviceID = [DeviceIDRotator getCurrentDeviceID];
    return [deviceID UTF8String];
}

File: DeviceIDRotator/DeviceIDRotator.swift
import Foundation
import UIKit

@objc public class DeviceIDRotator: NSObject {
    
    private static let deviceIDKey = "com.swiggy.customDeviceID"
    private static let shared = DeviceIDRotator()
    
    @objc public static func sharedInstance() -> DeviceIDRotator {
        return shared
    }
    
    /// Get the current custom device ID or generate a new one
    @objc public func getDeviceID() -> String {
        if let storedID = UserDefaults.standard.string(forKey: DeviceIDRotator.deviceIDKey) {
            return storedID
        }
        return generateNewDeviceID()
    }
    
    /// Generate a new UUID-based device ID
    @objc public func generateNewDeviceID() -> String {
        let newID = UUID().uuidString.lowercased()
        UserDefaults.standard.set(newID, forKey: DeviceIDRotator.deviceIDKey)
        UserDefaults.standard.synchronize()
        return newID
    }
    
    /// Rotate to a new device ID
    @objc public func rotateDeviceID() -> String {
        return generateNewDeviceID()
    }
    
    /// Get current device ID without generating new one
    @objc public func getCurrentDeviceID() -> String? {
        return UserDefaults.standard.string(forKey: DeviceIDRotator.deviceIDKey)
    }
    
    /// Clear stored device ID (will generate new on next get)
    @objc public func clearDeviceID() {
        UserDefaults.standard.removeObject(forKey: DeviceIDRotator.deviceIDKey)
        UserDefaults.standard.synchronize()
    }
}

// Hook for UIDevice.identifierForVendor
extension UIDevice {
    
    @objc static func swizzleIdentifierForVendor() {
        guard self == UIDevice.self else { return }
        
        let originalSelector = #selector(getter: UIDevice.identifierForVendor)
        let swizzledSelector = #selector(UIDevice.swizzled_identifierForVendor)
        
        guard let originalMethod = class_getInstanceMethod(self, originalSelector),
              let swizzledMethod = class_getInstanceMethod(self, swizzledSelector) else {
            return
        }
        
        let didAddMethod = class_addMethod(self,
                                          originalSelector,
                                          method_getImplementation(swizzledMethod),
                                          method_getTypeEncoding(swizzledMethod))
        
        if didAddMethod {
            class_replaceMethod(self,
                              swizzledSelector,
                              method_getImplementation(originalMethod),
                              method_getTypeEncoding(originalMethod))
        } else {
            method_exchangeImplementations(originalMethod, swizzledMethod)
        }
    }
    
    @objc func swizzled_identifierForVendor() -> UUID? {
        // Return custom device ID as UUID
        let customID = DeviceIDRotator.sharedInstance().getDeviceID()
        return UUID(uuidString: customID)
    }
}

// Hook for ASIdentifierManager (Advertising Identifier)
@objc class ASIdentifierManagerHook: NSObject {
    
    @objc static func swizzleAdvertisingIdentifier() {
        // This would hook ASIdentifierManager if available
        // Note: Requires linking against AdSupport framework
    }
    
    @objc static func getAdvertisingIdentifier() -> String {
        return DeviceIDRotator.sharedInstance().getDeviceID()
    }
}
